ocos/kernels/sentencepiece_tokenizer.cc

```diff
@@ -0,0 +1,110 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "sentencepiece_processor.h"
+#include "sentencepiece_tokenizer.hpp"
+
+KernelSentencepieceTokenizer::KernelSentencepieceTokenizer(OrtApi api) : BaseKernel(api) {
+}
+
+static void _check_dimension_constant(Ort::CustomOpApi ort, const OrtValue* ort_value, const char* name) {
+  OrtTensorDimensions dimensions(ort, ort_value);
+  if (dimensions.size() != 1 || dimensions[0] != 1)
+    throw std::runtime_error(MakeString(
+        name, " must contain only one element. It has ", dimensions.size(), " dimensions."));
+}
+
+void KernelSentencepieceTokenizer::Compute(OrtKernelContext* context) {
+  // Setup inputs
+  const OrtValue* ort_model = ort_.KernelContext_GetInput(context, 0);
+  const std::string* str_model = ort_.GetTensorData<std::string>(ort_model);
+  const OrtValue* ort_input = ort_.KernelContext_GetInput(context, 1);
+  const std::string* str_input = ort_.GetTensorData<std::string>(ort_input);
+  const OrtValue* ort_nbest_size = ort_.KernelContext_GetInput(context, 2);
+  const float* p_nbest_size = ort_.GetTensorData<float>(ort_nbest_size);
+  const OrtValue* ort_alpha = ort_.KernelContext_GetInput(context, 3);
+  const float* p_alpha = ort_.GetTensorData<float>(ort_alpha);
+  const OrtValue* ort_add_bos = ort_.KernelContext_GetInput(context, 4);
+  const bool* p_add_bos = ort_.GetTensorData<bool>(ort_add_bos);
+  const OrtValue* ort_add_eos = ort_.KernelContext_GetInput(context, 5);
+  const bool* p_add_eos = ort_.GetTensorData<bool>(ort_add_eos);
+  const OrtValue* ort_add_rev = ort_.KernelContext_GetInput(context, 6);
+  const bool* p_add_rev = ort_.GetTensorData<bool>(ort_add_rev);
+
+  // Verifications
+  _check_dimension_constant(ort_, ort_model, "model");
+  _check_dimension_constant(ort_, ort_nbest_size, "nbest_size");
+  _check_dimension_constant(ort_, ort_alpha, "alpha");
+  _check_dimension_constant(ort_, ort_add_bos, "add_bos");
+  _check_dimension_constant(ort_, ort_add_eos, "add_eos");
+  _check_dimension_constant(ort_, ort_add_rev, "add_rev");
```

Comment:
Maybe the inputs of this op are a bit much. Could we put some inputs to attribute, such as `model`?

Reply:
It would be more efficient but the current C API of onnxruntime does not allow that. I created an issue on onnxruntime to track that (https://github.com/microsoft/onnxruntime/issues/6409).